File: PriceCalculator/Rounder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PriceCalculator
{
    internal class Rounder
    {
        /// <summary>
        /// A sorted array of tuples each containing two doubles, the first (and that upon which the list is sorted) a value to compare inputs against and the second a value to round items less than the first value to.
        /// </summary>
        public Tuple<double, double>[] RoundingThresholds { get; private set; }
        /// <summary>
        /// <para>The percentage of the value the input is rounded to which determines whether the input should be rounded up or down.</para>
        /// <para>Defaults to 0.3</para>
        /// </summary>
        public double RoundDownThreshold { get; private set; }
        public double Round(double input)
        {
            // Find the appropriate threshold
            Tuple<double, double> threshold = new Tuple<double,double>(0.0, 0.5);
            for (int i = 0; i < RoundingThresholds.Length; i++)
            {
                threshold = RoundingThresholds[i];
                if (threshold.Item1 > input) break;
            }

            // Round
            if ((input % threshold.Item2) / threshold.Item2 < RoundDownThreshold)
            {
                return input - (input % threshold.Item2);
            }
            return input + (threshold.Item2 - (input % threshold.Item2));
        }

        public Rounder(Tuple<double, double>[] roundingThresholds, double roundDownThreshold = .3)
        {
            if (roundingThresholds.Length < 1) throw new ArgumentException("At least one rounding threshold must be defined");
            if (roundDownThreshold <= 0 | roundDownThreshold >= 1) roundDownThreshold = .3;
            RoundingThresholds = roundingThresholds;
            RoundDownThreshold = roundDownThreshold;
        }
    }
}

File: rounder.py


class Rounder():
  def round(self, input):
    threshold = (0, 0.5)
    #  find the appropriate threshold
    for t in self.rounding_thresholds:
      threshold = t
      if t[0] > input:
        break
    
    # round
    if (round((input % threshold[1]) / threshold[1], 5) <= self.round_down_threshold):  # added a round operator to accomodate floating point error
      return input - (input % threshold[1])
    return input + (threshold[1] - (input % threshold[1]))

  def __init__(self, _rounding_thresholds=[(0.0, 0.5)], _round_down_threshold = 0.3) -> None:
    self.rounding_thresholds = _rounding_thresholds
    self.round_down_threshold = _round_down_threshold
    pass


if __name__ == '__main__':
  # tests
  r = Rounder(_rounding_thresholds=[(5, .5), (20, 2), (100, 13)], _round_down_threshold=0.2)
  assert(r.round(1.3) == 1.5)
  assert(r.round(1.1) == 1)
  assert(r.round(.9) == 1)

  assert(r.round(10.3) == 10)
  assert(r.round(10.51) == 12)
  assert(r.round(19.9) == 20)

  assert(r.round(20.1) == 26)
  assert(r.round(35.2) == 39)
  assert(r.round(40.25) == 39)
  print('All tests passed')